### ISSUE TITLE
Consider alternative for keycloak-gatekeeper used in FAPI-RW conformance test run environment

### ISSUE THREAD

**Issue by tnorimat** [keycloak/kc-sig-fapi#46]
The conformance test run environment uses keycloak-gatekeeper.

According to https://groups.google.com/forum/#!topic/keycloak-dev/Jb5iqo3kHXY , the keycloak-gatekeeper and its successor Louketo becomes EOL. Therefore, an alternative for keycloak-gatekeeper is needed in the long run.

It is helpful if someone propose this alternative and replace the current keycloak-gatekeeper with it.

**Comment by kannan-ra** [comment]
I would certainly consider these - 
* apache 
* nginx

I know Hans Zandbelt @zmartzone whose [apache oidc module](https://github.com/zmartzone/mod_auth_openidc
) is OpenID certified &  
[supports keycloak integration](https://github.com/zmartzone/mod_auth_openidc/wiki/Keycloak)

**Comment by VinodAnandan** [comment]
This may be solved with using Keycloak's userinfo endpoint. Do we know if the Keycloak's userinfo endpoint can support FPAI headers ( e.g: x-fapi-interactionid)? Joseph also suggested deploying a FAPI compliant resource server.

**Comment by zandbelt** [comment]
As the author of [mod_auth_openidc](https://github.com/zmartzone/mod_auth_openidc) I would like to express my interest to help out on the way forward using Apache as suggested by @kannan-ra . Making mod_auth_openidc FAPI compliant has been on the backlog for a while now and concrete demand would help to push it forward.

**Comment by VinodAnandan** [comment]
@zandbelt Thank you very much for your interest to help out. We definitely need more community help :) . Is it okay if I assign this task to you ?

**Comment by zandbelt** [comment]
I will investigate first and let you know about my plans in more detail

**Comment by zandbelt** [comment]
I believe mod_auth_openidc can be configured to be FAPI 1.0 RO compliant as it is today: the value of the header `x-fapi-interaction-id` can be echo-ed in the response using mod_headers. When using version >=2.3.9 and combining it with [mod_tokenbinding](https://github.com/zmartzone/mod_token_binding#application) it should be FAPI 1.0 RW compliant. Let me know if I have overseen something.

I'll leave it up to this group to create the actual setup to this issue's purpose, but I'd be happy to provide pointers.

**Comment by wadahiro** [comment]
@tnorimat @VinodAnandan @kannan-ra I think I can help to implement switching to mod_auth_openidc or nginx (lua-resty-openidc). Could you assign me?

@zandbelt According to the spec, not only echo `x-fapi-interaction-id`, it must return generated UUID value if the request header was not provided. Is it possible to setup mod_headers for that?

> shall set the response header x-fapi-interaction-id to the value received from the corresponding fapi client request header or to a RFC4122 UUID value if the request header was not provided to track the interaction

**Comment by VinodAnandan** [comment]
Thank you @wadahiro. I have assigned it to you. Should we consider a minimal docker image with Apache+required modules?

**Comment by wadahiro** [comment]
> Should we consider a minimal docker image with Apache+required modules?

@VinodAnandan Yes. Since @zandbelt pushed [fapi branch](https://github.com/zmartzone/mod_auth_openidc/tree/fapi) for the apache module which supports `x-fapi-interaction-id`, I try to create a docker image as a new gateway for our resource server instead of keycloak-gatekeeper image.

**Comment by wadahiro** [comment]
I've tried [fapi branch](https://github.com/zmartzone/mod_auth_openidc/tree/fapi) today, but it seems that using `ProxyPass` directive doesn't return `x-fapi-interaction-id` header (It was ok when returning static html in httpd server).

Also, I tried to implement it using [zmartzone/lua-resty-openidc](https://github.com/zmartzone/lua-resty-openidc) and some lua scripts. it passed a basic test `fapi-rw-id2` which is primarily 'happy' flows of `fapi-rw-id2-test-plan`. 😄
So, I will continue to test with Nginx and make a pull request if there is no problem.
